### DISCUSSION TITLE
support regex rule

### DISCUSSION THREAD

**Discussion by aptend** [aptend/typing-transformer-obsidian#71] [Ideas]
#70 

something like `(r/.+) -> [$1](https://reddit.com/$1) `

**Comment by YousufSSyed** [comment]
> https://github.com/aptend/typing-transformer-obsidian/issues/70#issuecomment-1711048276
> I'll reiterate, it could be done after a word / phrase is completed when a space is typed after it, and or when a sentence is completed.

**Comment by caasion** [comment]
A potential workaround for the performance issue is to be able to define a "start" and "trigger" for the regex matching. This approach is also used by many other plugins or software. Think about discord: you have to type a colon `:` in order to trigger the emoji selection and you have to press `enter` to select the emoji. Since regex can get crazily big, it would be a good idea to control it like so.

Using the same regex, `(r/.+)`, let `%=` be the "start" and `enter` be the trigger.

Typing `%=` would start detecting the regex. These characters would disappear and only turn into a visual indicator that you are now in "regex auto replace mode". Now, if you type `r/abc` then press `enter`, then you will get the regex replacement.

I think this is a good start for implementing regex in this plugin. However, this feature should be introduced as an experimental feature because there may be many bugs and edge cases.

If you are concerned about the friction of typing the "start" characters, there could also be an option you can turn on "regex auto replace mode" to check everything you are typing against the regex queries the user has defined.

**Reply by YousufSSyed** [reply]
Would you be able to just check all the characters a user started typing since they opened the note or switched tabs and use a space as a trigger?

**Reply by YousufSSyed** [reply]
Have you considered asking LLMs like ChatGPT, or SO how to implement this?

**Reply by caasion** [reply]
> Would you be able to just check all the characters a user started typing since they opened the note or switched tabs and use a space as a trigger?

The "regex auto replace mode" would have such a function, but I think testing will be needed to see whether if scanning for regex since opening the note or switching tabs would be viable. As for the trigger, you would be able to set any character or maybe hotkey as one.

> Have you considered asking LLMs like ChatGPT, or SO how to implement this?

I'm not really the one who does the coding (though I am trying to learn :smile:). Right now, we are just brainstorming about what we want and how we can do things.

**Reply by YousufSSyed** [reply]
> I'm not really the one who does the coding (though I am trying to learn 😄). Right now, we are just brainstorming about what we want and how we can do things.

I need to pay more attention to usernames, I just thought it was you because I glanced that your pfp has similar colors 😂.

**Comment by YousufSSyed** [comment]
@aptend Have you looked at Espanso? https://github.com/espanso/espanso/, https://espanso.org/. It has regex support: https://espanso.org/docs/matches/regex-triggers/.

While I'm can use this on Mac, I'd like an Obsidian plugin that could do the same so it also works on mobile.

**Reply by aptend** [reply]
OK, I'll check that. Thank you for the information!

**Comment by joluj** [comment]
Regex supports tokens that can help define a proper matches, also for ending a phrase. This can be very hard to understand if you are not familiar with regex, but with the help of ChatGPT and especially some examples, it should be relatively easy to define proper rules.

Example Rule: `'(r/.+)(\W)' -> '[$1](https://reddit.com/$1)$2'`

Breakdown:
- `r/` matches the string `r/`.
- `.+` matches any characters, but at least one.
- `\W` matches any non-word character. This ensures that the rule is only triggered when the phrase has ended, e.g. by punctuation marks or a line break.
- The round brackets `(`...`)` create groups. The first group is the phrase itself (e.g. `r/obsidian`). The second group is the character that ends the actual phrase. It needs to be appended as well.

Example Replacement: `r/obsidian.` -> `[r/obsidian](https://reddit.com/r/obsidian).` (note the period that is included at the end!).

Example to play around with: https://regex101.com/r/GjOn7i/1

---

I think an example is definitely needed for users to understand this pattern. The simple example from my comment should be enough for most users build their own patterns.

Implementing it that way would not require any further UI changes and thus would feel like the normal non-regex replacements.

**Comment by jeansordes** [comment]
As an alternative, there is the plugin "Linter" that can apply custom regex to a document automatically (either when you save, or when the file changes, it's just an option)

@YousufSSyed For your use case I think it could do the trick
